cairo-libs/packages/cometbft/src/verifier.cairo

```diff
@@ -0,0 +1,308 @@
+use cometbft::errors::CometErrors;
+use cometbft::types::{
+    BlockIdFlag, Header, NonAbsentCommitVotes, NonAbsentCommitVotesTrait, Options, SignedHeader,
+    SimpleValidator, TrustedBlockState, UntrustedBlockState, UntrustedBlockStateTrait, ValidatorSet,
+    ValidatorSetTrait, VotingPowerTally, VotingPowerTallyTrait,
+};
+use cometbft::utils::{Fraction, TWO_THIRDS};
+use core::num::traits::CheckedAdd;
+use protobuf::primitives::array::{ByteArrayAsProtoMessage, BytesAsProtoMessage};
+use protobuf::primitives::numeric::U64AsProtoMessage;
+use protobuf::types::message::ProtoCodecImpl;
+use protobuf::types::wkt::{Duration, Timestamp};
+use crate::utils::{MerkleHashImpl, u32_8_to_array_u8};
+
+/// Verifies an update header received as the `header` field of `MsgUpdateClient`.
+///
+/// NOTE: The checks included here are the same as those in
+/// [`tendermint-light-client-verifier`](https://github.com/informalsystems/tendermint-rs/blob/6cc391c80ae88615508bd87285571ba130b604c/light-client-verifier/src/verifier.rs#L268-L24).
+pub fn verify_update_header(
+    untrusted: UntrustedBlockState, trusted: TrustedBlockState, options: Options, now: Timestamp,
+) {
+    verify_validator_sets(@untrusted);
+    validate_against_trusted(@untrusted, @trusted, @options, now);
+    verify_header_is_from_past(@untrusted, @options, now);
+    verify_commit_against_trusted(untrusted, trusted, options);
+}
+
+/// Verifies a misbehaviour header received as the `header` field of `MsgUpdateClient`.
+///
+/// NOTE: The checks included here are the same as those in
+/// [`tendermint-light-client-verifier`](https://github.com/informalsystems/tendermint-rs/blob/6cc31c80ae88615508bd87285571ba130b604c/light-client-verifier/src/verifier.rs#L310-L312)
+pub fn verify_misbehaviour_header(
+    untrusted: UntrustedBlockState, trusted: TrustedBlockState, options: Options, now: Timestamp,
+) {
+    verify_validator_sets(@untrusted);
+    validate_against_trusted(@untrusted, @trusted, @options, now);
+    verify_commit_against_trusted(untrusted, trusted, options)
+}
+
+pub fn verify_validator_sets(untrusted: @UntrustedBlockState) {
+    validator_sets_match(untrusted.validators, untrusted.signed_header.header.validators_hash);
+    // validator_sets_match(
+    //     untrusted.next_validators, untrusted.signed_header.header.next_validators_hash,
+    // );
```

Comment:
Why don't we check next validators?

Reply:
next validator field is removed.

check https://github.com/informalsystems/ibc-starknet/pull/366/files#r2022718374